// ng-jhipster/src/interceptor/interceptable-http.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/*
 Copyright 2013-2017 the original author or authors from the JHipster project.

 This file is part of the JHipster project, see https://jhipster.github.io/
 for more information.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
import { JhiHttpInterceptor } from './http.interceptor';
import { forwardRef, Inject, Injectable } from '@angular/core';
import { Http, ConnectionBackend, RequestOptions, RequestMethod, Headers } from '@angular/http';
var JhiInterceptableHttp = (function (_super) {
    __extends(JhiInterceptableHttp, _super);
    function JhiInterceptableHttp(backend, defaultOptions, interceptors // see the issue generator-jhipster#4794
    ) {
        var _this = _super.call(this, backend, defaultOptions) || this;
        /**
                 * building a responsibility chain of http interceptors, so when processXXXInterception is called on first interceptor,
                 * all http interceptors are called in a row
                 * Note: the array of interceptors are wired in customHttpProvider of the generated Jhipster app in file `http.provider.ts`
                 *
                */
        if (interceptors && interceptors.length > 0) {
            interceptors.reduce(function (chain, current) {
                chain.successor = current;
                return current;
            });
            _this.firstInterceptor = interceptors[0];
        }
        return _this;
    }
    JhiInterceptableHttp.prototype.request = function (url, options) {
        // Response interceptor needs to be called only once after the final request here
        // Every HTTP method will go through this request method
        return this.interceptResponse(_super.prototype.request.call(this, url, options));
    };
    JhiInterceptableHttp.prototype.get = function (url, options) {
        return _super.prototype.get.call(this, url, this.interceptRequest(url, RequestMethod.Get, options));
    };
    JhiInterceptableHttp.prototype.post = function (url, body, options) {
        return _super.prototype.post.call(this, url, body, this.interceptRequest(url, RequestMethod.Post, options, body));
    };
    JhiInterceptableHttp.prototype.put = function (url, body, options) {
        return _super.prototype.put.call(this, url, body, this.interceptRequest(url, RequestMethod.Put, options, body));
    };
    JhiInterceptableHttp.prototype.delete = function (url, options) {
        return _super.prototype.delete.call(this, url, this.interceptRequest(url, RequestMethod.Delete, options));
    };
    JhiInterceptableHttp.prototype.patch = function (url, body, options) {
        return _super.prototype.patch.call(this, url, body, this.interceptRequest(url, RequestMethod.Patch, options, body));
    };
    JhiInterceptableHttp.prototype.head = function (url, options) {
        return _super.prototype.head.call(this, url, this.interceptRequest(url, RequestMethod.Head, options));
    };
    JhiInterceptableHttp.prototype.options = function (url, options) {
        return _super.prototype.options.call(this, url, this.interceptRequest(url, RequestMethod.Options, options));
    };
    JhiInterceptableHttp.prototype.interceptRequest = function (url, method, options, body) {
        if (!options) {
            options = new RequestOptions();
        }
        if (!options.headers) {
            options.headers = new Headers();
        }
        options.url = options.url || url;
        options.method = options.method || method;
        options.body = options.body || body;
        return !this.firstInterceptor ? options : this.firstInterceptor.processRequestInterception(options);
    };
    JhiInterceptableHttp.prototype.interceptResponse = function (observable) {
        return !this.firstInterceptor ? observable : this.firstInterceptor.processResponseInterception(observable);
    };
    JhiInterceptableHttp.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    JhiInterceptableHttp.ctorParameters = function () { return [
        { type: ConnectionBackend, },
        { type: RequestOptions, },
        { type: Array, decorators: [{ type: Inject, args: [forwardRef(function () { return JhiHttpInterceptor; }),] },] },
    ]; };
    return JhiInterceptableHttp;
}(Http));
export { JhiInterceptableHttp };
